### PR TITLE
[RW-2602][risk=no] Implementation of new CB universal search

### PR DESCRIPTION
**Pull request by freemabd**
CB universal search will allow researchers to search for source or standard criteria/concepts within/without a hierarchy. Researchers will be able to switch back and forth between source/standard. **Invision** 
- https://projects.invisionapp.com/d/main/default/?origin=v7#/console/12820961/356869861/preview?newCollabSignupFlow=0

This PR addresses all API work for implementing the new CB universal search. This code is currently feature flagged and is turned off until the implementation of: 
- https://precisionmedicineinitiative.atlassian.net/browse/RW-2682

In order to test RW-2682 with the new API we will have to turn off elastic temporarily. When in progress we can work follow-up tickets to fix elastic search. 
- https://precisionmedicineinitiative.atlassian.net/browse/RW-2870
- https://precisionmedicineinitiative.atlassian.net/browse/RW-2872
- https://precisionmedicineinitiative.atlassian.net/browse/RW-2875

This mainly addresses 2 things:
1. Refactor of all BQ query builders into a single query builder class called 
**SearchGroupItemQueryBuilder**.
2. Refactor lookups of all parent criteria between the CB query builder and elastic search filters. This allows for better maintainability when feature changes happen. 

This PR will deprecate the following classes:
- AbstractQueryBuilder
- CodesQueryBuilder
- DemoQueryBuilder
- DrugQueryBuilder
- MeasurementQueryBuilder
- PMQueryBuilder
- PPIQueryBuilder
- VisitQueryBuilder
- FactoryKey
- QueryBuilderFactory

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+    // Otherwise build sql against flat denormalized search table
+    for (SearchParameter param : searchGroupItem.getSearchParameters()) {
+      if (param.getAttributes().isEmpty()) {
+        childConceptIds.addAll(childConceptIds(criteriaLookup, ImmutableList.of(param)));
+        // make sure we only add the concept ids sql template once
+        if (!queryParts.contains(CONCEPT_ID_IN)) {
+          queryParts.add(CONCEPT_ID_IN);
+        }
+      } else {
+        StringBuilder bpSql = new StringBuilder(BP_SQL);
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java]: I'm a bit lost here - why do we go straight into a blood pressure query? At this point we just have a searchGroupItem with attributes listed

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+    // create the source/standard flag named parameter
+    String standardParam =
+        addQueryParameterValue(queryParams, QueryParameterValue.int64(isStandard ? 1 : 0));
+    // format the base sql with all query parts
+    String baseSql = String.format(BASE_SQL, standardParam) + queryPartsSql;
+    // build modifier sql if modifiers exists
+    String modifiedSql = buildModifierSql(baseSql, queryParams, searchGroupItem.getModifiers());
+    // split off the sql conditions in case were building a temporal sql statement
+    // otherwise this is just ignored
+    String conditionsSql = baseSql.split("where")[1] + "%s";
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java]: This is a bit tough to follow and verify - perhaps the conditions clause can be constructed separately? Seems difficult to verify that there's exactly one WHERE clause in the sql query.

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+    // build the inner temporal sql if this search group item is temporal
+    // otherwise return modifiedSql
+    return buildInnerTemporalQuery(
+        modifiedSql, conditionsSql, queryParams, searchGroupItem.getModifiers(), mention);
+  }
+
+  /** Build sql statement for demographics */
+  private static String buildDemoSql(
+      Map<String, QueryParameterValue> queryParams, SearchGroupItem searchGroupItem) {
+    // The UI implementation of demographics only allows for 1 search parameter per search group
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java]: Should probably codify anything else as a BadRequest

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+    // item
+    SearchParameter param = searchGroupItem.getSearchParameters().get(0);
+    switch (CriteriaType.valueOf(param.getType())) {
+      case AGE:
+        // age attribute should always have 2 operands since the UI uses a slider
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java]: nit: In terms of separation of concerns, the API shouldn't really be depending on UI implementation details.

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+   */
+  private static String getAgeDateAndEncounterSql(
+      Map<String, QueryParameterValue> queryParams, List<Modifier> modifiers) {
+    List<Modifier> ageDateAndEncounterModifiers = new ArrayList<>();
+    ageDateAndEncounterModifiers.add(getModifier(modifiers, ModifierType.AGE_AT_EVENT));
+    ageDateAndEncounterModifiers.add(getModifier(modifiers, ModifierType.EVENT_DATE));
+    ageDateAndEncounterModifiers.add(getModifier(modifiers, ModifierType.ENCOUNTERS));
+    StringBuilder modifierSql = new StringBuilder();
+    for (Modifier modifier : ageDateAndEncounterModifiers) {
+      if (modifier != null) {
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java]: nit: to reduce nesting, I would invert this and continue

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java (hunk 2 above)
Comment (freemabd) [reply api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java]: Fixed.

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java (hunk 3 above)
Comment (freemabd) [reply api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java]: Done.

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java (hunk 4 above)
Comment (freemabd) [reply api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java]: Done.

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java (hunk 5 above)
Comment (freemabd) [reply api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java]: Done.

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java (hunk 1 above)
Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java]: So `attribute.conceptId` - the **only** use case for this is blood pressure? If so, possibly add this to the comment at L180

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java [hunk 6]
```diff
@@ -0,0 +1,583 @@
...
+      queryPartsSql = String.format(queryPartsSql, conceptIdsParam);
+    }
+    // format the base sql with all query parts
+    String baseSql = BASE_SQL + queryPartsSql;
+    // build modifier sql if modifiers exists
+    String modifiedSql = buildModifierSql(baseSql, queryParams, searchGroupItem.getModifiers());
+    // build the inner temporal sql if this search group item is temporal
+    // otherwise return modifiedSql
+    return buildInnerTemporalQuery(
+        modifiedSql, queryPartsSql + "%s", queryParams, searchGroupItem.getModifiers(), mention);
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java]: Could you add a comment to explain this `%s`?

File: api/src/test/java/org/pmiops/workbench/cohortbuilder/querybuilder/util/CriteriaLookupUtilTest.java [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+            .conceptId(21600002L);
+    SearchRequest searchRequest =
+        new SearchRequest()
+            .addIncludesItem(
+                new SearchGroup()
+                    .addItemsItem(new SearchGroupItem().addSearchParametersItem(searchParameter)));
+    assertEquals(
+        ImmutableMap.of(searchParameter, new HashSet<>(childConceptIds)),
+        lookupUtil.buildCriteriaLookupMap(searchRequest));
+    cbCriteriaDao.delete(drugNode1.getId());
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/cohortbuilder/querybuilder/util/CriteriaLookupUtilTest.java]: DirtiesContext? https://github.com/all-of-us/workbench/blob/dffa11812709f42266c6dfab4c106b43bc5838c6/api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java#L97

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java (hunk 1 above)
Comment (freemabd) [reply api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java]: Yes that is true. Comment added.

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java (hunk 6 above)
Comment (freemabd) [reply api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java]: I removed it, not needed.

File: api/src/test/java/org/pmiops/workbench/cohortbuilder/querybuilder/util/CriteriaLookupUtilTest.java (hunk 7 above)
Comment (freemabd) [reply api/src/test/java/org/pmiops/workbench/cohortbuilder/querybuilder/util/CriteriaLookupUtilTest.java]: Added @DirtiesContext and removed the manual deletes.

**Pull request merged**